### PR TITLE
db_bench: Enhancement - Support multiple benchmarks groups, each with its own config (#220)

### PR DESCRIPTION
**Pull request by udi-speedb**

### CODE REVIEW

File: tools/db_bench_tool.cc [hunk 1]
```diff
@@ -8657,34 +8698,107 @@ void ValidateMetadataCacheOptions() {
...
+  std::string fs_uri{kInvalidString};
+  bool simulate_hdd{false};
+  std::string simulate_hybrid_fs_file{kInvalidString};
+  int32_t simulate_hybrid_hdd_multipliers{-1};
+  int64_t seed{-1};
+};
+
+FirstGroupApplicableFlags first_group_applicable_flags;
+
+void RecordFirstGroupApplicableFlags() {
```
Comment (Yuval-Ariel) [inline tools/db_bench_tool.cc:8738]: why are only these flags recorded and checked? where are they later used which could cause a problem?

Comment (Yuval-Ariel) [reply tools/db_bench_tool.cc:8738]: for example, the FLAGS_env_uri is only used for the first group in ValidateAndProcessEnvFlags so i dont see a reason to call ValidateSubsequentGroupsDoNotOverrideApplicableFlags on it.

**Review by Yuval-Ariel — CHANGES_REQUESTED** [review]

File: tools/db_bench_tool.cc [hunk 2]
```diff
@@ -110,6 +110,10 @@ using GFLAGS_NAMESPACE::SetUsageMessage;
 #define IF_ROCKSDB_LITE(Then, Else) Else
 #endif
 
+// This is NOT a gflags flag but is defined as is only so it is displayed in the
+// gflags help (generated automatically)
```
Comment (Yuval-Ariel) [inline tools/db_bench_tool.cc:114]: can you rephrase this so its a bit more clear?

File: tools/db_bench_tool.cc [hunk 3]
```diff
@@ -1285,6 +1289,8 @@ DEFINE_string(fs_uri, "",
               " with --env_uri."
               " Creates a default environment with the specified filesystem.");
 #endif  // ROCKSDB_LITE
+// TODO - It seems the following 3 flags need to be inside the '#ifndef
+// ROCKSDB_LITE' macros
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:1293]: OK. I will open a separate issue for this.

File: tools/db_bench_tool.cc [hunk 4]
```diff
@@ -8827,18 +8979,155 @@ int db_bench_tool(int argc, char** argv) {
...
+//
+// Note that all the groups after the first one assume you are using the same DB
+// (they enforce FLAGS_use_existing_db=true) => the tool will abort execution if
+// the user requests to run a benchmark that requires a fresh db (e.g.,
+// requesting the 'fillseq' benchmark)
+//
+// Return Value:
+// 0 If all of the groups completed successfully or an error reported by the
+// runner of the failed group (subsequent groups will NOT be run).
+//
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:9059]: OK. I will take applicable parts and duplicate in the help section

File: tools/db_bench_tool.cc [hunk 5]
```diff
@@ -8827,18 +8979,155 @@ int db_bench_tool(int argc, char** argv) {
...
+//    configuration. The first group (the master group) sets the initial
+//    configuration
+//     for all subsequent groups. Subsequent groups may override the initial
+//     configuration.
+//
+//    Important: The DB-s are also opened when the first group runs. When one
+//    group completes and the next starts,
+//     the db-s are retained (they are kept open). However, the DB options are
+//     set when the DB-s are opened. Therefore, attempts to override options in
+//     subsequent groups are silently ignored.
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:9026]: Verifying that all of the options have remained as is is too complex (there are tons of options and comparing them is complicated).
I will emphasize in the help message that overriding them  in groups > 1 is SILENTLY ignored

File: tools/db_bench_tool.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:1293]: @Yuval-Ariel I have opened [https://github.com/speedb-io/speedb/issues/256](url)
Will remove my comment

File: tools/db_bench_tool.cc (hunk 1 above)
Comment (udi-speedb) [reply tools/db_bench_tool.cc:8738]: I have explicitly checked those flags that I am ignoring in groups > 1.

File: tools/db_bench_tool.cc (hunk 2 above)
Comment (udi-speedb) [reply tools/db_bench_tool.cc:114]: Done

**Comment by udi-speedb** [comment]
@Yuval-Ariel - I have pushed a commit with the CR updates we have discussed.
Coming next are additional changes that I accidentally performed on #221 but belong here.

File: tools/db_bench_tool.cc [hunk 6]
```diff
@@ -110,9 +110,33 @@ using GFLAGS_NAMESPACE::SetUsageMessage;
...
 
-// This is NOT a gflags flag but is defined as is only so it is displayed in the
-// gflags help (generated automatically)
-DEFINE_string(groups, "", "Run db_bench in benchmark groups mode.");
+// The groups flags is NOT a standard GFLAGS flag. It is a special flag that is
+// used to indicate that the tool is run in a multiple-groups mode (see the help
+// description for the flag for more details). It is defined using GFLAGS
+// definition syntax so it is included in GFLAGS' automatic help generation.
+DEFINE_string(
+    groups, "",
```
Comment (Yuval-Ariel) [inline tools/db_bench_tool.cc:118]: i dont see an instruction to put the -groups flags in the first place after ./db_bench (2nd arg total) here and its definitely mandatory. it is in the Synthax part but i think a more clear instruction to place the flag in the 2nd position is required.

File: tools/db_bench_tool.cc [hunk 7]
```diff
@@ -9012,48 +9034,54 @@ int db_bench_tool_run_group(int group_num, int num_groups, int argc,
...
-// the user requests to run a benchmark that requires a fresh db (e.g.,
-// requesting the 'fillseq' benchmark)
+// Notes:
+// 1. The DB-s are opened when the master group runs. When one group completes
+//     and the next starts,
+//     the db-s are retained (they are kept open). However, the DB options are
+//     set only when the DB-s are opened. Therefore, attempts to override
+//     options in subsequent groups are SILENTLY ignored.
+// 2. A consequence of (1) is that an attempt to run benchmarks that require a
+//     fres db in groups > 1
```
Comment (Yuval-Ariel) [inline tools/db_bench_tool.cc:9078]: *fresh

File: tools/db_bench_tool.cc [hunk 8]
```diff
@@ -9012,48 +9034,54 @@ int db_bench_tool_run_group(int group_num, int num_groups, int argc,
 //
 // There are 2 modes of operation:
 // 1. Single-group: The tool is run with a set of flags once, running all
-//    specified benchmarks and exiting. This is the default mode.
+//    specified benchmarks and exiting. This is the DEFULT mode.
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:8940]: Yes. Fixed

File: tools/db_bench_tool.cc [hunk 9]
```diff
@@ -110,9 +110,33 @@ using GFLAGS_NAMESPACE::SetUsageMessage;
...
+// used to indicate that the tool is run in a multiple-groups mode (see the help
+// description for the flag for more details). It is defined using GFLAGS
+// definition syntax so it is included in GFLAGS' automatic help generation.
+DEFINE_string(
+    groups, "",
+    "Run db_bench in benchmark groups mode (The default is single-group mode). "
+    "In this mode benchmarks "
+    "are grouped, "
+    "and each group has its own configuration (values for the "
+    "supported flags). "
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:148]: The group's configuration is dictated by the values of its supported flags.
Please propose a clarification / rephrasing

File: tools/db_bench_tool.cc (hunk 6 above)
Comment (udi-speedb) [reply tools/db_bench_tool.cc:143]: Added.

File: tools/db_bench_tool.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:148]: Removed the sentence in parenthesis since it's confusing

File: tools/db_bench_tool.cc (hunk 7 above)
Comment (udi-speedb) [reply tools/db_bench_tool.cc:8981]: Fixed

File: tools/db_bench_tool.cc [hunk 10]
```diff
@@ -110,9 +110,33 @@ using GFLAGS_NAMESPACE::SetUsageMessage;
...
+    "The first group is the MASTER group. This group sets the "
+    "initial configuration "
+    "for all subsequent groups. Subsequent groups may override the "
```
[the remark this answers is not in the document]
Comment (udi-speedb) [reply tools/db_bench_tool.cc:151]: I have reformatted.

**Review by Yuval-Ariel — APPROVED** [review]

**Comment by udi-speedb** [comment]
@Yuval-Ariel - Thanks for the review.
I will now rebase on latest main and squash the commits.
